libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -226,13 +203,26 @@ onError g m r k e = liftIO $ do
 
 -- | Log an 'Error' response for debugging purposes.
 logError :: MonadIO m => Logger -> Maybe Request -> Error -> m ()
-logError g r (Error c l m) = liftIO $ Log.debug g logMsg
+logError g mr (Error c l m) = liftIO $ Log.debug g logMsg
   where
     logMsg = field "code" (statusCode c)
            . field "label" l
-           . field "request" (fromMaybe "N/A" (lookupRequestId =<< r))
+           . field "request" (maybe "N/A" reqinfo mr)
            . msg (val "\"" +++ m +++ val "\"")
 
+    -- Using unsafePerformIO here is important because we only want it to run if log level is
+    -- set to 'Debug'.
+    reqinfo r = show ( lookupRequestId r
+                     , r
+                       -- show the bytestring containing the request body, this way it can be
+                       -- read back into ghci for forensics.
+                     , IOUnsafe.unsafePerformIO $ lazyRequestBody r
```

Comment:
I'd prefer `unsafeInterleaveIO` for things like this I think.

It still runs in regular IO; but defers the actual work until the value is forced (which presumably won't happen if we're at a different logging level); I think it should work here 😄

Reply:
I'm interested!  I would have expected this `unsafePerformIO` to be only reachable if the message is to be printed, and the `show` above is called on the tuple containing it.  How can execution of `IO` be triggered if the call to `unsafePerformIO` is deep inside an unevaluated thunk?

Reply:
Also, `unsafeInterleaveIO` has the wrong type.  How can it replace `unsafePerformIO`?  Or should I use the former inside the latter?